colossus/src/main/scala/colossus/protocols/http/HttpResponse.scala

```diff
@@ -135,7 +135,14 @@ trait HttpResponseBuilding {
   def initialVersion: HttpVersion
 
   def respond[T: HttpBodyEncoder](code: HttpCode, data: T, headers: HttpHeaders = HttpHeaders.Empty) = {
-    HttpResponse(HttpResponseHead(initialVersion, code, headers), HttpBody(data))
+    HttpResponse(HttpResponseHead(initialVersion, code, headersWithContentType[T](headers)), HttpBody(data))
+  }
+
+  def headersWithContentType[T: HttpBodyEncoder](headers: HttpHeaders = HttpHeaders.Empty): HttpHeaders = {
+    implicitly[HttpBodyEncoder[T]].contentType match {
```

Comment:
Not sure this is going to work. What happens if you test:

```
HttpResponse.ok("hello").withContentType("other")
```

Reply:
IMO this is a separate issue / outside the scope of this PR, as this is fixing the current architecture

you can also do
```
HttpResponse.ok("Hello", HttpHeaders(HttpHeader("Content-Type", "first"))).withContentType("second")
```
to end up with multiple content-type's in the response header

Reply:
I agree this is kind of part of a larger issue, since you can have the same problem with other headers we generate like `Date` and `Content-Length`.  

Hopefully it should be clear that by requiring the body type to have a `HttpBodyEncoder` the response will automatically have the right content-type, but we should definitely make sure though to have clear documentation about this.  Maybe we should add scaladoc to `respond` noting this.

Reply:
opened an issue to move the conversation out of this PR: https://github.com/tumblr/colossus/issues/649